<a href="https://colab.research.google.com/github/ValeriiLubiako/MyFirstRepository/blob/master/Python_Analytics_Sem003_HA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Основы языка Python для аналитиков (семинары)#
##Урок 3. Изменение таблиц в Pandas##

###Условие 1: Задача 1###
_Скачать данные c домами._

_Считать данные с помощью pandas_

__Вывести на экран первые 5 строк__

1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю <br>
стоимость за один кв. метр общей площади

1.2 Создать новый признак delta_renovated, который будет содержать разницу в <br>годах между годом реновацией дома и годом постройки дома. Если реновации <br> дома не было, то в новом признаке поставьте 0.

1.3 Создайте признаки года продажи, месяца продажи

1.4 Удалите признаки date, zipcode, lat, long

In [ ]:
import pandas as pd

In [ ]:
from google.colab import files
uploaded=files.upload()

Saving kc_house_data.csv to kc_house_data.csv


In [ ]:
df=pd.read_csv('kc_house_data.csv', encoding='latin-1')

Вывести на экран первые 5 строк

In [ ]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю
стоимость за один кв. метр общей площади

In [ ]:
df['price_per_sq_lot'] = df['price']/df['sqft_lot']

In [ ]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  price_per_sq_lot  
0        5650         39.274336  
1        7639         74.288870  
2        8062         18.000000  
3        5000        120.800000  
4        7503         63.118812  

[5 rows x 22 columns]

1.2. Создать новый признак delta_renovated, который будет содержать разницу в
годах между<br> годом реновацией дома и годом постройки дома.<br> Если реновации
дома не было, то в новом признаке поставьте 0.

In [ ]:
import numpy as np

In [ ]:
df['delta_renovated'] = np.where(df['yr_renovated'] == 0, 0, df['yr_renovated']-df['yr_built'])

In [ ]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
0      5650     1.0           0     0  ...              0      1955   
1      7242     2.0           0     0  ...            400      1951   
2     10000     1.0           0     0  ...              0      1933   
3      5000     1.0           0     0  ...            910      1965   
4      8080     1.0           0     0  ...              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
0             0    98178  47.5112 -122.257           1340        5650   
1          1991    98125  47.7210 -122.319           1690        7639   
2             0    98028  47.7379 -122.233           2720        8062   
3             0    98136  47.5208 -122.393           1360        5000   
4             0    98074  47.6168 -122.045           1800        7503   

   price_per_sq_lot  delta_renovated  
0         39.274336                0  
1         74.288870               40  
2         18.000000                0  
3        120.800000                0  
4         63.118812                0  

[5 rows x 23 columns]

1.3 Создайте признаки года продажи, месяца продажи

In [ ]:
df['yr_sold'] = df['date'][0:3]

In [ ]:
df['yr_sold'] = df['date'].apply(lambda x: x[0:4])

In [ ]:
df['yr_sold'] = df['date'].str[0:3]

In [ ]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   
2 -122.233           2720        8062         18.000000                0   
3 -122.393           1360        5000        120.800000                0   
4 -122.045           1800        7503         63.118812                0   

           yr_sold  year  
0  20141013T000000  2014  
1  20141209T000000  2014  
2  20150225T000000  2015  
3              NaN  2014  
4              NaN  2015  

[5 rows x 25 columns]

In [ ]:
df['yr_sold'] = df['date'].apply(partline)

In [ ]:
def partline(my_str:str):
  my_str = my_str[0:4]
  return my_str

In [ ]:
df['yr_sold'] = df['date'].apply(partline)

In [ ]:
df['year'] = df['date'].str[0:4]

In [ ]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  zipcode      lat     long  \
0      5650     1.0           0     0  ...    98178  47.5112 -122.257   
1      7242     2.0           0     0  ...    98125  47.7210 -122.319   
2     10000     1.0           0     0  ...    98028  47.7379 -122.233   
3      5000     1.0           0     0  ...    98136  47.5208 -122.393   
4      8080     1.0           0     0  ...    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0           1340        5650         39.274336                0   
1           1690        7639         74.288870               40   
2           2720        8062         18.000000                0   
3           1360        5000        120.800000                0   
4           1800        7503         63.118812                0   

           yr_sold  year  month_sold  
0  20141013T000000  2014          10  
1  20141209T000000  2014          12  
2  20150225T000000  2015          02  
3              NaN  2014          12  
4              NaN  2015          02  

[5 rows x 26 columns]

In [ ]:
def partline46(my_str:str):
  my_str = my_str[4:6]
  return my_str

In [ ]:
df['month_sold'] = df['date'].apply(partline46)

In [ ]:
df['month_sold'] = df['date'].apply(lambda x: x[4:6])

In [ ]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   
2 -122.233           2720        8062         18.000000                0   
3 -122.393           1360        5000        120.800000                0   
4 -122.045           1800        7503         63.118812                0   

   yr_sold  month_sold  
0     2014          10  
1     2014          12  
2     2015          02  
3     2014          12  
4     2015          02  

[5 rows x 25 columns]

1.4 Удалите признаки date, zipcode, lat, long

In [ ]:
df.drop(columns=['date', 'zipcode', 'lat',  'long'], inplace = True)



In [ ]:
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   
2           0     0          3  ...         770              0      1933   
3           0     0          5  ...        1050            910      1965   
4           0     0          3  ...        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650         39.274336                0   
1          1991           1690        7639         74.288870               40   
2             0           2720        8062         18.000000                0   
3             0           1360        5000        120.800000                0   
4             0           1800        7503         63.118812                0   

   yr_sold month_sold  
0     2014         10  
1     2014         12  
2     2015         02  
3     2014         12  
4     2015         02  

[5 rows x 21 columns]

##Условие 2: 2 задача##
Создайте датафрейм с клиентами:<br>
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],<br>
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})

2.1 Присоедините к таблице clients данные по домам через метод join

2.2 Присоедините к таблице clients данные по домам через метод merge

Это нужно, чтобы понимать, какие дома покупались клиентами
house_id - это индексы датафрейма с домами

In [ ]:
clients = pd.DataFrame({ 'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
                                       939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242, 2068,
                                       4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
                        'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207,
                                     6821600300, 7137950720, 9510920050, 6131600255, 5428000070, 1788800910,
                                     8100400160, 3123049142, 6306800010, 5083000375, 7920100025, 1951600150,
                                     809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205,
                                     3995700435, 8861700030, 3303980210, 7731100066, 8146100580, 825069097,
                                     3889100029, 9524100196]
                       })

In [ ]:
clients

client_id    house_id
0        1459  8965450190
1        4684  6823100225
2        3498  5104540330
3        3942  2131701075
4        4535  1522700060
5        2308  1189000207
6        2866  6821600300
7        2765  7137950720
8        1472  9510920050
9        4236  6131600255
10       2295  5428000070
11        939  1788800910
12       3840  8100400160
13        280  3123049142
14         20  6306800010
15       4332  5083000375
16       3475  7920100025
17       4213  1951600150
18       3113   809001400
19       4809   339600110
20       2134  1622049154
21       2242  1099600250
22       2068  8563000110
23       4929  2768100205
24       1384  3995700435
25       1589  8861700030
26       3317  3303980210
27       2260  7731100066
28       1727  8146100580
29       1764   825069097
30       1611  3889100029
31       1474  9524100196

In [ ]:
client_house_id = clients.set_index('house_id')
house_house_id = df.set_index('id')

In [ ]:
print(client_house_id.head())

            client_id
house_id             
8965450190       1459
6823100225       4684
5104540330       3498
2131701075       3942
1522700060       4535


In [ ]:
house_house_id.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
7129300520  221900.0         3       1.00         1180      5650     1.0   
6414100192  538000.0         3       2.25         2570      7242     2.0   
5631500400  180000.0         2       1.00          770     10000     1.0   
2487200875  604000.0         4       3.00         1960      5000     1.0   
1954400510  510000.0         3       2.00         1680      8080     1.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  \
id                                                                          
7129300520           0     0          3      7        1180              0   
6414100192           0     0          3      7        2170            400   
5631500400           0     0          3      6         770              0   
2487200875           0     0          5      7        1050            910   
1954400510           0     0          3      8        1680              0   

            yr_built  yr_renovated  sqft_living15  sqft_lot15  \
id                                                              
7129300520      1955             0           1340        5650   
6414100192      1951          1991           1690        7639   
5631500400      1933             0           2720        8062   
2487200875      1965             0           1360        5000   
1954400510      1987             0           1800        7503   

            price_per_sq_lot  delta_renovated yr_sold month_sold  
id                                                                
7129300520         39.274336                0    2014         10  
6414100192         74.288870               40    2014         12  
5631500400         18.000000                0    2015         02  
2487200875        120.800000                0    2014         12  
1954400510         63.118812                0    2015         02

In [ ]:
clients_houses_joined = client_house_id.join(house_house_id)

In [ ]:
clients_houses_joined.head()

client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
339600110        4809  395000.0         3       2.50         1610      3755   
809001400        3113  925000.0         3       1.00         1630      3200   
825069097        1764  770000.0         3       2.50         2650     40705   
1099600250       2242  260000.0         3       1.75         1710      6400   
1189000207       2308  387000.0         2       2.50         1170      1394   

            floors  waterfront  view  condition  ...  sqft_above  \
339600110      2.0           0     0          3  ...        1610   
809001400      1.5           0     0          4  ...        1630   
825069097      2.0           0     0          3  ...        2650   
1099600250     1.0           0     0          4  ...        1240   
1189000207     2.0           0     0          3  ...        1170   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
339600110               0      1987             0           1300        3823   
809001400               0      1912             0           1710        3600   
825069097               0      1994             0           2550       42625   
1099600250            470      1976             0           1600        6400   
1189000207              0      2001             0           1250        3136   

            price_per_sq_lot  delta_renovated  yr_sold month_sold  
339600110         105.193076                0     2014         09  
809001400         289.062500                0     2014         09  
825069097          18.916595                0     2014         06  
1099600250         40.625000                0     2014         12  
1189000207        277.618364                0     2014         10  

[5 rows x 21 columns]

In [ ]:
clients_houses_merged = client_house_id.merge(df, left_on='house_id', right_on='id').set_index('client_id')

In [ ]:
clients_houses_merged.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
client_id                                                                     
1459       8965450190  295000.0         3       2.50         1500      3060   
4684       6823100225  700000.0         4       1.75         1870      6000   
3498       5104540330  679000.0         4       2.50         3680      7236   
3942       2131701075  420000.0         3       1.75         1720      5000   
4535       1522700060  518000.0         4       2.75         2520     14021   

           floors  waterfront  view  condition  ...  sqft_above  \
client_id                                       ...               
1459          2.0           0     0          3  ...        1500   
4684          1.0           0     0          5  ...        1670   
3498          2.0           0     0          3  ...        3680   
3942          1.5           0     0          3  ...        1720   
4535          2.0           0     0          3  ...        2520   

           sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
client_id                                                                     
1459                   0      1994             0           2700        7734   
4684                 200      1949             0           1710        6000   
3498                   0      2006             0           3310        7180   
3942                   0      1932          2009           1410        8300   
4535                   0      1999             0           2330       14007   

           price_per_sq_lot  delta_renovated  yr_sold month_sold  
client_id                                                         
1459              96.405229                0     2015         02  
4684             116.666667                0     2015         04  
3498              93.836374                0     2015         05  
3942              84.000000               77     2014         12  
4535              36.944583                0     2014         06  

[5 rows x 21 columns]

##Условие 3: 3 задача##
<b><i>Составьте несколько сводных таблиц</i></b><br>
3.1 Найдите среднюю стоимость домов в зависимости от количества спален

Отсортируйте от меньшей стоимости к большей

3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

3.4 Каких домов в зависимости от этажности и количества спален больше?

3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома


In [ ]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price')

price
bedrooms              
1         3.176429e+05
2         4.013727e+05
0         4.095038e+05
3         4.662321e+05
11        5.200000e+05
4         6.354195e+05
33        6.400000e+05
5         7.865998e+05
10        8.193333e+05
6         8.255206e+05
9         8.939998e+05
7         9.511847e+05
8         1.105077e+06

Минимальная, средняя и максимальная стоимости домов в зависимости от состояния

In [ ]:
df.groupby('condition').agg({'price': ['min', 'mean', 'max']})

price                          
                min           mean        max
condition                                    
1           78000.0  334431.666667  1500000.0
2           80000.0  327287.145349  2555000.0
3           75000.0  542012.578148  7062500.0
4           89000.0  521200.390033  7700000.0
5          110000.0  612418.089359  3650000.0

 Таблица с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [ ]:
df.pivot_table(
    index='waterfront',
    columns='view',
    aggfunc='count',
    values='id'
)

view              0      1      2      3      4
waterfront                                     
0           19489.0  331.0  955.0  491.0  184.0
1               NaN    1.0    8.0   19.0  135.0

In [ ]:
df1 = df.pivot_table(
    index='floors',
    columns='bedrooms',
    aggfunc='count',
    values='id'
)
df1

bedrooms   0      1       2       3       4      5      6     7    8    9   \
floors                                                                       
1.0       4.0  162.0  1951.0  5455.0  2383.0  605.0  104.0   9.0  5.0  NaN   
1.5       NaN   21.0   182.0   786.0   698.0  185.0   30.0   7.0  1.0  NaN   
2.0       6.0   12.0   497.0  3118.0  3682.0  775.0  119.0  19.0  6.0  4.0   
2.5       NaN    1.0     5.0    56.0    58.0   23.0   14.0   2.0  NaN  2.0   
3.0       2.0    3.0   123.0   405.0    61.0   13.0    5.0   1.0  NaN  NaN   
3.5       1.0    NaN     2.0     4.0     NaN    NaN    NaN   NaN  1.0  NaN   

bedrooms   10   11   33  
floors                   
1.0       1.0  NaN  1.0  
1.5       NaN  NaN  NaN  
2.0       2.0  1.0  NaN  
2.5       NaN  NaN  NaN  
3.0       NaN  NaN  NaN  
3.5       NaN  NaN  NaN

Максимальное количество составляют 1-этажные дома с 3-мя спальнями - 5455.

Медианная стоимость домов в зависимости от состояния дома и оценки дома

In [ ]:
pd.crosstab(
    index=df['condition'],
    columns=df['grade'],
    aggfunc='median',
    values=df['price'],
).fillna('0')

grade            1         3         4         5         6         7   \
condition                                                               
1          142000.0         0  150000.0  190000.0  255000.0  403500.0   
2                 0  280000.0  145000.0  180000.0  235000.0  305000.0   
3                 0   75000.0  205000.0  234475.0  265000.0  357500.0   
4                 0         0  238525.0  229975.0  295000.0  390000.0   
5                 0  262000.0         0  227450.0  285475.0  456000.0   

grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0          0          0          0          0          0  
2          429000.0   715000.0  1752500.0          0          0          0  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0          0

Медианная стоимость домов в зависимости от состояния дома и оценки дома